db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -715,58 +697,92 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
     }
     s = RunManualCompaction(cfd, ColumnFamilyData::kCompactAllLevels,
                             final_output_level, options, begin, end, exclusive,
-                            false, max_file_num_to_ignore);
+                            false, port::kMaxUint64);
   } else {
-    for (int level = 0; level <= max_level_with_files; level++) {
-      int output_level;
-      // in case the compaction is universal or if we're compacting the
-      // bottom-most level, the output level will be the same as input one.
-      // level 0 can never be the bottommost level (i.e. if all files are in
-      // level 0, we will compact to level 1)
-      if (cfd->ioptions()->compaction_style == kCompactionStyleUniversal ||
-          cfd->ioptions()->compaction_style == kCompactionStyleFIFO) {
-        output_level = level;
-      } else if (level == max_level_with_files && level > 0) {
-        if (options.bottommost_level_compaction ==
-            BottommostLevelCompaction::kSkip) {
-          // Skip bottommost level compaction
-          continue;
-        } else if (options.bottommost_level_compaction ==
-                       BottommostLevelCompaction::kIfHaveCompactionFilter &&
-                   cfd->ioptions()->compaction_filter == nullptr &&
-                   cfd->ioptions()->compaction_filter_factory == nullptr) {
-          // Skip bottommost level compaction since we don't have a compaction
-          // filter
-          continue;
+    int first_overlapped_level = kInvalidLevel;
+    int max_overlapped_level = kInvalidLevel;
+    // max_file_num_to_ignore can be used to filter out newly created SST files,
+    // useful for bottom level compaction in a manual compaction
+    uint64_t max_file_num_to_ignore = port::kMaxUint64;
+    uint64_t next_file_number = port::kMaxUint64;
+    {
+      InstrumentedMutexLock l(&mutex_);
+      Version* base = cfd->current();
+      for (int level = 0; level < base->storage_info()->num_non_empty_levels();
+           level++) {
+        bool overlap = true;
+        if (begin != nullptr && end != nullptr) {
+          s = base->OverlapWithLevelIterator(ReadOptions(), file_options_,
+                                             *begin, *end, level, &overlap);
```

Comment:
Here we are still inside DB Mutex (held from line 709), but this operation can be very expensive and involve disk I/O. We should do it outside the DB mutex. Maybe hold a reference of SuperVersion and do it use the Version pointed by the super version.